Parser/pegen/pegen.c

```diff
@@ -264,66 +268,21 @@ raise_tokenizer_init_error(PyObject *filename)
 }
 
 static inline PyObject *
-get_error_line(char *buffer)
-{
-    char *newline = strchr(buffer, '\n');
-    if (newline) {
-        return PyUnicode_FromStringAndSize(buffer, newline - buffer);
-    }
-    else {
-        return PyUnicode_FromString(buffer);
-    }
-}
-
-static int
-tokenizer_error_with_col_offset(Parser *p, PyObject *errtype, const char *errmsg)
+get_error_line(char *buffer, int is_file)
 {
-    PyObject *errstr = NULL;
-    PyObject *value = NULL;
-    size_t col_number = -1;
-
-    errstr = PyUnicode_FromString(errmsg);
-    if (!errstr) {
-        return -1;
-    }
-
-    PyObject *loc = NULL;
-    if (p->start_rule == Py_file_input) {
-        loc = PyErr_ProgramTextObject(p->tok->filename, p->tok->lineno);
-    }
-    if (!loc) {
-        loc = get_error_line(p->tok->buf);
+    const char *newline;
+    if (is_file) {
+        newline = strrchr(buffer, '\n');
+    } else {
+        newline = strchr(buffer, '\n');
```

Comment:
I don't really understand why this is needed.

Reply:
On eval or single input cases (or fstring), it needs to search for until the first newline. But on file input, it needs to include until the last one.

Reply:
`get_error_line` will only get called with `is_file=1` if there was a memory error in `PyErr_ProgramTextObject`. Also, `p->tok->buf` only contains the line of the error, so it won't have more than one newlines (there might be corner cases to this though). I think that this might add a bit of unnecessary complexity. Are there test cases that fail without this?

Reply:
This [test](https://github.com/python/cpython/blob/master/Lib/test/test_exceptions.py#L221-L230) fails with this message
```
0:00:00 load avg: 0.58 Run tests sequentially
0:00:00 load avg: 0.58 [1/1] test_exceptions
test test_exceptions failed -- Traceback (most recent call last):
  File "/home/isidentical/cpython/cpython/Lib/test/test_exceptions.py", line 221, in testSyntaxErrorOffset
    check(b"""if 1:
  File "/home/isidentical/cpython/cpython/Lib/test/test_exceptions.py", line 190, in check
    self.assertIn(line, cm.exception.text)
AssertionError: "                '''quux'''" not found in "                '''"

test_exceptions failed

== Tests result: FAILURE ==

1 test failed:
    test_exceptions

Total duration: 170 ms
Tests result: FAILURE
```

Reply:
What if we always use `strrchr`? Would that work?

Reply:
Unfortunately not, there are some usages that depend on it when used with eval/exec. Check error raising part on codeop maybe compile function.

Reply:
Understood! Thanks a lot for your patience in explaining everything.